cpp/src/io/fst/logical_stack.cuh

```diff
@@ -0,0 +1,530 @@
+
+    // We propagate the write from the LHS if it's a write
+    bool is_lhs_write = (lhs != read_symbol);
+
+    return (is_rhs_read && is_lhs_write) ? lhs : rhs;
+  }
+
+  /// The read_symbol that is supposed to be replaced
+  StackSymbolT read_symbol;
+};
+
+/**
+ * @brief Helper function object class to convert a StackOp to the stack symbol of that
+ * StackOp.
+ */
+struct StackOpToStackSymbol {
+  template <typename StackLevelT, typename ValueT>
+  constexpr CUDF_HOST_DEVICE ValueT operator()(StackOp<StackLevelT, ValueT> const& kv_op) const
+  {
+    return kv_op.value;
+  }
+};
+
+/**
+ * @brief Replaces all operations that apply to stack level '0' with the empty stack symbol
+ */
+template <typename StackOpT>
+struct RemapEmptyStack {
+  constexpr CUDF_HOST_DEVICE StackOpT operator()(StackOpT const& kv_op) const
+  {
+    return kv_op.stack_level == 0 ? empty_stack_symbol : kv_op;
+  }
+  StackOpT empty_stack_symbol;
+};
+
+/**
+ * @brief Function object to return only the stack_level part from a StackOp instance.
+ */
+struct StackOpToStackLevel {
+  template <typename StackLevelT, typename ValueT>
+  constexpr CUDF_HOST_DEVICE StackLevelT operator()(StackOp<StackLevelT, ValueT> const& kv_op) const
+  {
+    return kv_op.stack_level;
+  }
+};
+
+/**
+ * @brief Retrieves an iterator that returns only the `stack_level` part from a StackOp iterator.
+ */
+template <typename StackOpItT>
+auto make_stack_level_iterator(StackOpItT it)
+{
+  return thrust::make_transform_iterator(it, StackOpToStackLevel{});
+}
+
+/**
+ * @brief Retrieves an iterator that returns only the `value` part from a StackOp iterator.
+ */
+template <typename StackOpItT>
+auto make_value_iterator(StackOpItT it)
+{
+  return thrust::make_transform_iterator(it, StackOpToStackSymbol{});
+}
+
+}  // namespace detail
+
+/**
+ * @brief Takes a sparse representation of a sequence of stack operations that either push something
+ * onto the stack or pop something from the stack and resolves the symbol that is on top of the
+ * stack.
+ *
+ * @tparam StackLevelT Signed integer type that must be sufficient to cover [-max_stack_level,
+ * max_stack_level] for the given sequence of stack operations. Must be signed as it needs to cover
+ * the stack level of any arbitrary subsequence of stack operations.
+ * @tparam StackSymbolItT An input iterator type that provides the sequence of symbols that
+ * represent stack operations
+ * @tparam SymbolPositionT The index that this stack operation is supposed to apply to
+ * @tparam StackSymbolToStackOpTypeT Function object class to transform items from StackSymbolItT to
+ * stack_op_type
+ * @tparam TopOfStackOutItT Output iterator type to which StackSymbolT are being assigned
+ * @tparam StackSymbolT The internal type being used (usually corresponding to StackSymbolItT's
+ * value_type)
+ * @tparam OffsetT Signed or unsigned integer type large enough to index into both the sparse input
+ * sequence and the top-of-stack output sequence
```

Comment:
Please align text for each tag (`@`). Otherwise the doxygen is very difficult to read.

Reply:
I understand your preference for readability. Following the discussion from  https://github.com/rapidsai/cudf/pull/11137#discussion_r911532945, however, I'm inclined to leave this as is for now